src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs

```diff
@@ -0,0 +1,557 @@
+            protected override BoundStatement GenerateStateMachineCreation(LocalSymbol stateMachineVariable, NamedTypeSymbol frameType)
+            {
+                // PROTOTYPE(async-streams): TODO review this (what is this error case at the start?)
+
+                // If the async method's result type is a type parameter of the method, then the AsyncTaskMethodBuilder<T>
+                // needs to use the method's type parameters inside the rewritten method body. All other methods generated
+                // during async rewriting are members of the synthesized state machine struct, and use the type parameters
+                // structs type parameters.
+                AsyncMethodBuilderMemberCollection methodScopeAsyncMethodBuilderMemberCollection;
+                if (!AsyncMethodBuilderMemberCollection.TryCreate(F, method, null, out methodScopeAsyncMethodBuilderMemberCollection))
+                {
+                    return new BoundBadStatement(F.Syntax, ImmutableArray<BoundNode>.Empty, hasErrors: true);
+                }
+
+                var bodyBuilder = ArrayBuilder<BoundStatement>.GetInstance();
+                LocalSymbol builderVariable = F.SynthesizedLocal(methodScopeAsyncMethodBuilderMemberCollection.BuilderType, null);
+
+                // local.$builder = System.Runtime.CompilerServices.AsyncTaskMethodBuilder<typeArgs>.Create();
+                bodyBuilder.Add(
+                    F.Assignment(
+                        F.Field(F.Local(stateMachineVariable), _builderField.AsMember(frameType)),
+                        F.StaticCall(
+                            null,
+                            methodScopeAsyncMethodBuilderMemberCollection.CreateBuilder)));
+
+                // local.$stateField = NotStartedStateMachine;
+                bodyBuilder.Add(
+                    F.Assignment(
+                        F.Field(F.Local(stateMachineVariable), stateField.AsMember(frameType)),
+                        F.Literal(StateMachineStates.NotStartedStateMachine)));
+
+                // builder = local.$stateField.builder;
+                bodyBuilder.Add(
+                    F.Assignment(
+                        F.Local(builderVariable),
+                        F.Field(F.Local(stateMachineVariable), _builderField.AsMember(frameType))));
+
+                // local._valueOrEndPromise = new ManualResetValueTaskSourceLogic<bool>(stateMachine);
+                MethodSymbol mrvtslCtor =
+                    F.WellKnownMethod(WellKnownMember.System_Threading_Tasks_ManualResetValueTaskSourceLogic_T__ctor)
+                    .AsMember((NamedTypeSymbol)_promiseOfValueOrEndField.Type);
+
+                bodyBuilder.Add(
+                    F.Assignment(
+                        F.Field(F.Local(stateMachineVariable), _promiseOfValueOrEndField.AsMember(frameType)),
+                        F.New(mrvtslCtor, F.Local(stateMachineVariable))));
+
+                // PROTOTYPE(async-streams): Why do we need AsMember?
```

Comment:
>Why do we need AsMember [](start = 45, length = 23)

Is it needed to emit the correct references to `_promiseIsActive` within the `MoveNext` for a generic async method?